openupgradelib/openupgrade_120.py

```diff
@@ -0,0 +1,259 @@
+    _r(selector=".pagination > li > a", class_add="page-link"),
+
+    # Breadcrumbs
+    _r(selector=".breadcrumb > li", class_add="breadcrumb-item"),
+
+    # Labels and badges
+    _r("label", "badge"),
+    _r("badge-default", "badge-secondary"),
+    *(_r("label-%s" % c3, "badge-%s" % c4) for (c3, c4) in CONTEXTS),
+
+    # Convert panels, thumbnails and wells to cards
+    _r("panel", "card"),
+    _r("panel-body", "card-body"),
+    _r("panel-default panel-group"),
+    _r("panel-footer", "card-footer"),
+    _r("panel-heading", "card-header"),
+    _r("panel-title", "card-title"),
+    *(_r("panel-%s" % c3, "bg-%s" % c4) for (c3, c4) in CONTEXTS),
+    _r("well", "card card-body"),
+    _r("thumbnail", "card card-body"),
+
+    # Progress
+    *(_r("progress-bar-%s" % c3, "bg-%s" % c4) for (c3, c4) in CONTEXTS),
+    _r("active", "progress-bar-animated", selector=".progress-bar.active"),
+
+    # Carousel
+    _r("carousel-control left", "carousel-control-prev"),
+    _r("carousel-control right", "carousel-control-next"),
+    _r("item", "carousel-item", selector=".carousel .item"),
+    _r("left", "carousel-item-left", selector=".carousel .left"),
+    _r("next", "carousel-item-next", selector=".carousel .next"),
+    _r("prev", "carousel-item-prev", selector=".carousel .prev"),
+    _r("right", "carousel-item-right", selector=".carousel .right"),
+
+    # Utilities
+    *(_r("hidden%s" % t3, "d-none%s" % t4) for (t3, t4) in TIERS),
+    _r("hidden-print", "d-print-none"),
+    *(_r("visible%s-%s" % (t3, vis), "d-")
+      for (t3, t4), vis in product(TIERS, BS3_VISIBLES)),
+    *(_r("visible-print-%s" % vis, "d-print-%s" % vis)
+      for vis in BS3_VISIBLES),
+    _r("pull-left", "float-left"),
+    _r("pull-right", "float-right"),
+)
+
+
+def convert_bootstrap_3to4(html_string, pretty_print=True):
+    """Convert an HTML string from Bootstrap 3 to 4.
+
+    :param str html_string:
+        Raw HTML fragment to convert.
+
+    :param bool pretty_print:
+        Indicate if you wish to return the HTML pretty formatted.
+
+    :return str:
+        Raw HTML fragment converted.
+    """
+    return convert_xml_fragment(html_string, BS4_REPLACEMENTS, pretty_print)
+
+
+def convert_field_bootstrap_3to4(env, model_name, field_name, domain=None,
+                                 method='orm'):
+    """This converts all the values for the given model and field, being
+    able to restrict to a domain of affected records.
+
+    :param env: Odoo environment.
+    :param model_name: Name of the model that contains the field.
+    :param field_name: Name of the field that contains the BS3 HTML content.
+    :param domain: Optional domain for filtering records in the model
+    :param method: 'orm' (default) for using ORM; 'sql' for avoiding problems
+        with extra triggers in ORM.
+    """
+    if domain is None:
+        domain = []
+    Model = env[model_name]
+    recs = Model.with_context(active_test=False).search(domain)
```

Comment:
You're using ORM to search before checking if you have to use ORM. What if this is being executed in a pre script where the model is new and still doesn't exist in the `env`?

IMHO it would be better to provide a list of ids instead of a domain, so they can be browsed fine equally with or without orm.

If you agree, I can patch it. I still have local WIP changes.

Reply:
That can't happen, as you don't need to convert a non-existing field: you only convert fields of an existing model that are in BS3. Any way, you can add the SQL part for being totally coherent with the param `method`.

Reply:
I mean: imagine a module adds model X with field `html_description`. You can't find it through `Model = env[model_name]` because model X still doesn't exist in the pre stage. It would fail also when using raw SQL mode.

Reply:
The same explanation: you don't need to convert a non existing field, but as said, just unfold this search to be done in SQL when using `method='sql'`.

Reply:
I found easier to split in 2 methods.

Reply:
Well, 3 then: the main one and the 2 submethods, but I prefer this if. It's better for following the flow IMO.